### PR TITLE
✨ New command line option `--show-all` to display all connected devices and improved log messages

### PR DESCRIPTION
**Pull request by madchutney**
### Description

- Keep track of devices which could not be identified
- Add command line option to show all connected devices
- Generate log messages when targets cannot be identified
- Rename OnlineId attributes to match online database and API
- Fixed minor bug with HTM file reading

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [x]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

#### Problem connecting the online database

Auto mode:

```
▶ mbed-tools devices
ERROR: A problem occurred when looking up Mbed Targets for connected devices.
Increase the verbosity level to `-vvv` to see debug information. Use the `--traceback` argument to see a full stack trace.

▶ mbed-tools -v devices
WARNING: There was an error connecting to the online database. Please check your internet connection.
ERROR: A problem occurred when looking up Mbed Targets for connected devices.
Increase the verbosity level to `-vvv` to see debug information. Use the `--traceback` argument to see a full stack trace.

▶ mbed-tools -vv devices
INFO: Using the online database to identify Mbed Targets.
INFO: Unable to identify an Mbed Target using the offline database, trying the online database.
WARNING: There was an error connecting to the online database. Please check your internet connection.
ERROR: A problem occurred when looking up Mbed Targets for connected devices.
Increase the verbosity level to `-vvv` to see debug information. Use the `--traceback` argument to see a full stack trace.
```

Online Mode:

```
▶ mbed-tools devices
ERROR: A problem occurred when looking up Mbed Targets for connected devices.
Increase the verbosity level to `-vvv` to see debug information. Use the `--traceback` argument to see a full stack trace.

▶ mbed-tools -v devices
WARNING: There was an error connecting to the online database. Please check your internet connection.
ERROR: A problem occurred when looking up Mbed Targets for connected devices.
Increase the verbosity level to `-vvv` to see debug information. Use the `--traceback` argument to see a full stack trace.

▶ mbed-tools -vv devices
INFO: Using the online database (only) to identify Mbed Targets.
WARNING: There was an error connecting to the online database. Please check your internet connection.
ERROR: A problem occurred when looking up Mbed Targets for connected devices.

```

#### Device connected has a serial number which does not match a known Mbed Target (offline)

Standard:

```
▶ mbed-tools devices
No connected Mbed devices found.
```

Show all selected:

```
▶ mbed-tools devices --show-all
Board name    Serial number             Serial port             Mount point(s)    Build target(s)
------------  ------------------------  ----------------------  ----------------  -----------------
              02400201CCF63E703108C3C8  /dev/tty.usbmodem14202  /Volumes/MBED
```

Show all selected and verbose mode:

```
▶ mbed-tools -vv devices --show-all
INFO: Using the offline database (only) to identify Mbed Targets.
INFO: The device with the Serial Number: '02400201CCF63E703108C3C8' (Product Code: '0240') does not appear to be an Mbed Target .
Board name    Serial number             Serial port             Mount point(s)    Build target(s)
------------  ------------------------  ----------------------  ----------------  -----------------
              02400201CCF63E703108C3C8  /dev/tty.usbmodem14202  /Volumes/MBED
```

#### Device with the product code in the HTM file but no match (auto)

```
▶ mbed-tools devices
ERROR: Could not identify an Mbed Target with the Product Code: '0000'.
No connected Mbed devices found.

▶ mbed-tools -vv devices
INFO: Using the online database to identify Mbed Targets.
INFO: Unable to identify an Mbed Target using the offline database, trying the online database.
ERROR: Could not identify an Mbed Target with the Product Code: '0000'.
No connected Mbed devices found.
```

#### Device with a slug ande device type in the HTM file but no match (auto)

```
▶ mbed-tools -vv devices
▶ mbed-tools devices
ERROR: Could not identify an Mbed Target with the Slug: 'badger' and Target Type 'module'.
No connected Mbed devices found.

▶ mbed-tools -vv devices
INFO: Using the online database to identify Mbed Targets.
INFO: Unable to identify an Mbed Target using the offline database, trying the online database.
ERROR: Could not identify an Mbed Target with the Slug: 'badger' and Target Type 'module'.
No connected Mbed devices found.
```

### CODE REVIEW

File: tests/_internal/mbed_tools/test_list_connected_devices.py [hunk 1]
```diff
@@ -34,42 +34,99 @@ def test_informs_when_no_devices_are_connected(self, get_connected_devices):
...
+            Device, mbed_target=mock.create_autospec(MbedTarget, board_name="A"), serial_number="123"
+        )
+        device_2 = mock.create_autospec(
+            Device, mbed_target=mock.create_autospec(MbedTarget, board_name="B"), serial_number="456"
+        )
+        device_3 = mock.create_autospec(
+            Device, mbed_target=mock.create_autospec(MbedTarget, board_name="C"), serial_number="789"
+        )
 
         result = _sort_devices_by_name([device_3, device_1, device_2])
```
Comment (muchzill4) [inline tests/_internal/mbed_tools/test_list_connected_devices.py:96]: Should we rename `_sort_devices_by_name` to something more appropriate, now that sorting rules have changed?

File: mbed_devices/_internal/resolve_target.py [hunk 2]
```diff
@@ -84,5 +99,9 @@ def _get_all_htm_files_contents(directories: Iterable[pathlib.Path]) -> List[str
 
 
 def _is_htm_file(file: pathlib.Path) -> bool:
+    """Checks whether the file looks like an Mbed HTM file.
+
+    Note that if the file is manually manipulated the file may sometimes appears to be present but not be accessible.
+    """
     extensions = [".htm", ".HTM"]
-    return file.suffix in extensions and not file.name.startswith(".")
+    return file.suffix in extensions and not file.name.startswith(".") and os.access(file, os.R_OK)
```
Comment (muchzill4) [inline mbed_devices/_internal/resolve_target.py:107]: Could you add a test for this? How does `os.access(file, os.R_OK)` present itself? Could we not rely on exception instead?

Rationale: untested bugfix will hurt us in the future, code coverage lies.

File: mbed_devices/mbed_devices.py [hunk 3]
```diff
@@ -3,37 +3,53 @@
...
 
+from mbed_targets import MbedTarget
 from mbed_devices.device import Device
 from mbed_devices.exceptions import DeviceLookupFailed
 
 
-def get_connected_devices() -> Iterable[Device]:
-    """Returns Mbed Devices connected to host computer."""
-    devices = []
+def get_connected_devices() -> Tuple[List[Device], List[Device]]:
```
Comment (muchzill4) [inline mbed_devices/mbed_devices.py:19]: Personally, I'm not a fan of returning tuples. You need to remember to deal with two values and discarding via `_` is ugly. It makes sense when language supports pattern matching, but in python it just feels clunky.

I wonder if this wouldn't work better if `get_connected_devices` accepted an argument `include_unidentified` or such and always returned a list.

File: mbed_devices/_internal/resolve_target.py (hunk 2 above)
Comment (rwalton-arm) [reply mbed_devices/_internal/resolve_target.py:107]: I'd prefer we just caught an `OSError` here (if possible) instead of using `R_OK` as that's rather opaque.

File: mbed_devices/_internal/htm_file.py [hunk 4]
```diff
@@ -69,11 +69,13 @@
 class OnlineId(NamedTuple):
```
Comment (rwalton-arm) [inline mbed_devices/_internal/htm_file.py:69]: Not part of this PR, but I'd like to see this `NamedTuple` changed to a `dataclass(frozen=True)`

File: mbed_devices/_internal/resolve_target.py [hunk 5]
```diff
@@ -42,19 +47,29 @@ def resolve_target(candidate: CandidateDevice) -> MbedTarget:
...
     # Product code might be the first 4 characters of the serial number
     try:
-        return get_target_by_product_code(candidate.serial_number[:4])
+        product_code = candidate.serial_number[:4]
+        return get_target_by_product_code(product_code)
     except UnknownTarget:
+        # Most devices have a serial number so this may not be a problem
+        logger.info(
+            f"The device with the Serial Number: '{candidate.serial_number}' (Product Code: '{product_code}') "
+            f"does not appear to be an Mbed Target ."
```
Comment (rosiemiller2) [inline mbed_devices/_internal/resolve_target.py:71]:
```suggestion
            f"does not appear to be an Mbed Target."
```

File: mbed_devices/_internal/resolve_target.py [hunk 6]
```diff
@@ -42,19 +47,29 @@ def resolve_target(candidate: CandidateDevice) -> MbedTarget:
         try:
             return get_target_by_product_code(product_code)
         except UnknownTarget:
+            logger.error(f"Could not identify an Mbed Target with the Product Code: '{product_code}'.")
```
Comment (rwalton-arm) [inline mbed_devices/_internal/resolve_target.py:50]: Is it possible for this error to appear in relation to USB MSD's connected which aren't mbed enabled devices (i.e a memory stick)? Or is there some other filtering of candidates further down the stack which prevents that happening?

File: mbed_devices/mbed_devices.py [hunk 7]
```diff
@@ -3,37 +3,53 @@
...
+    """Returns Mbed Devices connected to host computer.
+
+    Returns:
+        A tuple containing two lists, the first containing devices identified as Mbed Targets the second those devices
+        connected but not identified as Mbed Targets.
+
+    Raises:
+        DeviceLookupFailed: TODO: find out when.
```
Comment (rosiemiller2) [inline mbed_devices/mbed_devices.py:27]: Is this a temporary todo before the pr is merged?

File: mbed_devices/_internal/mbed_tools/list_connected_devices.py [hunk 8]
```diff
@@ -17,9 +17,22 @@
...
+    is_flag=True,
+    default=False,
+    help="Show all connected devices, even those which are not Mbed Targets.",
+)
+def list_connected_devices(format: str, show_all: str) -> None:
     """Prints connected devices."""
-    devices = _sort_devices_by_name(get_connected_devices())
+    identified_devices, unidentified_devices = get_connected_devices()
+
+    if show_all:
```
Comment (rwalton-arm) [inline mbed_devices/_internal/mbed_tools/list_connected_devices.py:31]: I'm confused by this. `show_all` is a `str` yet we only check if it evaluates to `True`? Why not just use a `bool` flag? Maybe I'm missing something in the diff.

File: mbed_devices/_internal/resolve_target.py (hunk 2 above)
Comment (madchutney) [reply mbed_devices/_internal/resolve_target.py:107]: @muchzill4 I can add a test, it seems to manifest because of a problem with the file system. The directory is listed but not actually present. It may also occur if the file cannot be read due to permission, but I haven't tried this.

@rwalton-arm The exception occurs inside the list comprehension so there are two other options:
- Unpacking the creation of the list and catch the exception - the code is not so nice doing this
- Catch the error at a higher level and abort the read for all files is there is a read error - less robust

File: tests/_internal/mbed_tools/test_list_connected_devices.py (hunk 1 above)
Comment (madchutney) [reply tests/_internal/mbed_tools/test_list_connected_devices.py:96]: @muchzill4 Do you have a suggestion?

File: mbed_devices/mbed_devices.py (hunk 3 above)
Comment (madchutney) [reply mbed_devices/mbed_devices.py:19]: I'm not a fan either but there is not a use case where the unidentified are not returned.

File: mbed_devices/mbed_devices.py (hunk 7 above)
Comment (madchutney) [reply mbed_devices/mbed_devices.py:27]: Damn, I found out and forgot to update the comment. Ta

File: mbed_devices/mbed_devices.py (hunk 3 above)
Comment (rwalton-arm) [reply mbed_devices/mbed_devices.py:19]: If we have a fixed structure would it be clearer to return some kind of `Results` object with two attributes `identified` and `unidentified`? Then we side step the potentially subtle bugs that can be caused by tuple unpacking.

File: mbed_devices/_internal/resolve_target.py (hunk 2 above)
Comment (muchzill4) [reply mbed_devices/_internal/resolve_target.py:107]: How about:
```
try:
   return file.suffix in extensions and not file.name.startswith('.')
except SOME_ERROR:
   return False
```

I think that was the suggestion?

File: mbed_devices/mbed_devices.py (hunk 3 above)
Comment (madchutney) [reply mbed_devices/mbed_devices.py:19]: That's a good idea, I'll take a look at refactoring.

File: tests/_internal/mbed_tools/test_list_connected_devices.py (hunk 1 above)
Comment (muchzill4) [reply tests/_internal/mbed_tools/test_list_connected_devices.py:96]: `_sort_devices` would do I think. Thoughts?

File: mbed_devices/mbed_devices.py (hunk 3 above)
Comment (muchzill4) [reply mbed_devices/mbed_devices.py:19]: I don't follow. Are you saying we always want to return both? Isn't the cli handling 2 cases differently proving that there's a use case for `with_unindentified=True` and `with_unindentified=False`? 🤔 

I don't feel strongly about this, so 🤷‍♂ .

File: mbed_devices/_internal/resolve_target.py (hunk 6 above)
Comment (madchutney) [reply mbed_devices/_internal/resolve_target.py:50]: I ion't think as this information is extracted from the HTM file, USB MSB that have a serial number is later code and that just generates an info message and has different wording.

File: mbed_devices/_internal/mbed_tools/list_connected_devices.py (hunk 8 above)
Comment (madchutney) [reply mbed_devices/_internal/mbed_tools/list_connected_devices.py:31]: Good spot, it should be a bool as it is in the click options.

File: mbed_devices/_internal/resolve_target.py (hunk 2 above)
Comment (rwalton-arm) [reply mbed_devices/_internal/resolve_target.py:107]: That's what I was thinking of. I'm not a fan of using `os.access` as it can potentially give misleading results. The docs also mention the potential to create security holes as it's messing with UID/GID to verify access rights. Essentially it's a can of very complicated permissions related worms. Even the python docs recommend using EAFP techniques as suggested. I'd suspect there's also the possibly we would encounter subtle differences in behaviour on Windows, as it has an even more complicated permissions model.

Comment (madchutney) [reply mbed_devices/_internal/resolve_target.py:107]: @muchzill4 Unfortunately the exception doesn't occur there, it occurs in the list comprehension calling this function when the read is performed.

File: mbed_devices/mbed_devices.py (hunk 3 above)
Comment (madchutney) [reply mbed_devices/mbed_devices.py]: You're right it is, but take a look at the updated code and tell me what you think.

File: mbed_devices/_internal/htm_file.py (hunk 4 above)
Comment (madchutney) [reply mbed_devices/_internal/htm_file.py:69]: I'm not going to do this now, but that seems like a reasonable change.

File: tests/_internal/mbed_tools/test_list_connected_devices.py (hunk 1 above)
Comment (madchutney) [reply tests/_internal/mbed_tools/test_list_connected_devices.py]: Done

File: mbed_devices/_internal/resolve_target.py (hunk 2 above)
Comment (muchzill4) [reply mbed_devices/_internal/resolve_target.py:107]: Ah, I see now. I got confused by where the `os.access` check was put. For some reason I thought it happens when file is interrogated for its name parts, which makes _absolutely no sense_. 🤦‍♂ 

With that in mind, I think this fix should live closer to the origin of the problem (`file.read_text()` is where it happens, right?).

I think handling file readability does not belong in `_is_htm_file` but should instead be shifted to where comprehension happens. It'll probably no longer be a comprehension, but a `for` loop. We could also have an appropriate log message to make the debug logs even better!

What I'm thinking:

```
all_files = ...
all_htm_files = (file for file in all_files if _is_htm_file(file)
htm_files_contents = []
for htm_file in htm_files:
   try:
      html_files_contents.append(htm_file.read_text())
   except SOME_ERROR:
      logging.approriate_level(f"Can't read {htm_file}.")
return htm_files_contents
```

File: mbed_devices/mbed_devices.py (hunk 3 above)
Comment (muchzill4) [reply mbed_devices/mbed_devices.py]: 💯

File: mbed_devices/_internal/resolve_target.py (hunk 2 above)
Comment (madchutney) [reply mbed_devices/_internal/resolve_target.py:107]: Indeed, this is my option:

- Unpacking the creation of the list and catch the exception - the code is not so nice doing this

Comment (muchzill4) [reply mbed_devices/_internal/resolve_target.py:107]: Ah, that makes sense! I can only speak code snippets @madchutney . 😎

Comment (madchutney) [reply mbed_devices/_internal/resolve_target.py]: @muchzill4 I have done something similar to your suggestion but to facilitate testing I had to create a separate function. Not too bad although a lot more code.

File: tests/_internal/test_resolve_target.py [hunk 9]
```diff
@@ -119,3 +121,43 @@ def test_returns_contents_of_all_htm_files_in_given_directories(self):
...
+
+            result = _read_htm_file_contents([pathlib.Path("mbed.htm"), pathlib.Path("error.htm")])
+
+        self.assertEqual(result, ["foo"])
+
+
+class TestIsHtmFile(TestCase):
+    def test_lower_case_htm(self):
+        with Patcher() as patcher:
+            patcher.fs.create_file("mbed.htm", contents="foo")
```
Comment (muchzill4) [inline tests/_internal/test_resolve_target.py:139]: Correct me if I'm wrong, but since `_is_htm_file` operates on `pathlib.Path` only, I believe it's not necessary to patch the file system.

File: mbed_devices/_internal/resolve_target.py (hunk 2 above)
Comment (muchzill4) [reply mbed_devices/_internal/resolve_target.py]: Thanks. I agree it's more verbose. I think it's worth paying that price for better coverage.

File: tests/_internal/test_resolve_target.py (hunk 9 above)
Comment (madchutney) [reply tests/_internal/test_resolve_target.py]: I cannot correct you, these kind of evolved over time. Fixed now.
